contracts/src/batch-bridge/L2BatchBridgeGateway.sol

```diff
@@ -0,0 +1,250 @@
+    mapping(address => mapping(uint256 => bytes32)) public batchHashes;
+
+    /// @notice Mapping from token address to the amount of failed distribution.
+    mapping(address => uint256) public failedAmount;
+
+    /// @notice Mapping from batch hash to the distribute status.
+    mapping(bytes32 => bool) public isDistributed;
+
+    /*************
+     * Modifiers *
+     *************/
+
+    modifier onlyMessenger() {
+        if (_msgSender() != messenger) {
+            revert ErrorCallerNotMessenger();
+        }
+        _;
+    }
+
+    /***************
+     * Constructor *
+     ***************/
+
+    /// @param _counterpart The address of `L1BatchBridgeGateway` contract in L1.
+    /// @param _messenger The address of `L2ScrollMessenger` contract in L2.
+    constructor(address _counterpart, address _messenger) {
+        _disableInitializers();
+
+        counterpart = _counterpart;
+        messenger = _messenger;
+    }
+
+    /// @notice Initialize the storage of `L2BatchBridgeGateway`.
+    function initialize() external initializer {
+        __Context_init(); // from ContextUpgradeable
+        __ERC165_init(); // from ERC165Upgradeable
+        __AccessControl_init(); // from AccessControlUpgradeable
+        __AccessControlEnumerable_init(); // from AccessControlEnumerableUpgradeable
+
+        _grantRole(DEFAULT_ADMIN_ROLE, _msgSender());
+    }
+
+    /*****************************
+     * Public Mutating Functions *
+     *****************************/
+
+    /// @notice Receive batch bridged ETH from `L2ScrollMessenger`.
+    receive() external payable onlyMessenger {
+        // empty
+    }
+
+    /// @notice Finalize L1 initiated batch token deposit.
+    /// @param l1Token The address of the token in L1.
+    /// @param l2Token The address of the token in L2.
+    /// @param batchIndex The index of this batch bridge.
+    /// @param hash The hash of this batch.
+    function finalizeBatchDeposit(
+        address l1Token,
+        address l2Token,
+        uint256 batchIndex,
+        bytes32 hash
+    ) external onlyMessenger {
+        // this usually won't happen, check just in case.
+        if (tokenMapping[l2Token] != l1Token) {
+            revert ErrorL1TokenMismatched();
+        }
+        if (counterpart != IL2ScrollMessenger(messenger).xDomainMessageSender()) {
+            revert ErrorMessageSenderNotCounterpart();
+        }
+
+        batchHashes[l2Token][batchIndex] = hash;
+
+        emit FinalizeBatchDeposit(l1Token, l2Token, batchIndex);
+    }
+
+    /************************
+     * Restricted Functions *
+     ************************/
+
+    /// @notice Update layer 2 to layer 1 token mapping.
+    /// @param l2Token The address of corresponding ERC20 token on layer 2.
+    /// @param l1Token The address of ERC20 token on layer 1.
+    function updateTokenMapping(address l2Token, address l1Token) external onlyRole(DEFAULT_ADMIN_ROLE) {
```

Comment:
Why don't we read the mapping from `L2GatewayRouter`?

Reply:
yes, we can. Adding mapping here is to save the bridge gas.

Reply:
removed.